parallel_ohms_law: Added a new component.

malamast opened the pull request:
1. Uses parallel Ohm's law to calculate the parallel current
2. Then uses this parallel current to calculate the parallel electron velocity.
3. It can potentially help with numerical stability in 2D asymmetric simulations (Tokamak plasma edge sims) when drifts are included. 
4. This component requires collisions to be calculated first, so it should be called after collisions component.
5. When parallel_ohms_law is used then there is no need to evolve_momentum for electrons

/// Caution! This component is not carfully evaluated for multispecies cases with multiple ions.

/// Use parallel Ohm's law to calculate the parallel current. 
/// This law basically comes from the electron parallel momentum equation 
/// by neglecting inertial (small by m_e / m_i) and viscous terms 
/// (see Phys. Plasmas 10, 4744–4757 (2003) by Simakov & Catto).
/// Use the parallel current to calculate the electron parallel velocity
///
///   Jpar = - 1/eta * ∇phi + 1/(eta e n_e) * ∇P_e + (0.71 k_B)/(eta e) * ∇T_e  
///
/// where phi is the electric potential and eta is the resistivity.
///
/// After normalization the equation takes the form: 
///
///   Jpar = - ∇phi / eta  +  ∇P_e / (eta n_e) + 0.71 ∇T_e / eta 
///
/// where the normalization parameter for the resistivity eta is:
///
///   eta_norm = Tnorm / (rho_s0 * Nnorm * Cs0 * qe)
///
/// Then uses this parallel current to calculate the parallel electron velocity.
///
/// Note: This needs to be called after collisions and other
///       components which impose forces on electrons_

malamast committed:
parallel_ohms_law: Added a new component.

 1) Use parallel Ohm's law to calculate the parallel current

 2) then uses this parallel current to calculate the parallel electron velocity.

-This compoment requries cossisions to be calculated first so it should be called after collisions component.

-When parallel_ohms_law is used then there is no need to evolve_momentum for electrons

mikekryjak commented:
Hi @malamast, nice work. Can you tell me a bit more about how mean field drifts with this approach compare to what SOLPS, UEDGE etc are doing? It looks more or less like what I understand is in SOLPS, but I'm not a drifts expert and never looked at the equations in detail.

ZedThree commented on src/parallel_ohms_law.cxx:
```diff
@@ -0,0 +1,388 @@
+
+
+namespace {
+BoutReal clip(BoutReal value, BoutReal min, BoutReal max) {
```
This is unused and can be deleted

ZedThree commented on src/parallel_ohms_law.cxx:
```diff
@@ -0,0 +1,388 @@
+  }
+
+  eta = softFloor( resistivity_eta / eta_norm , resistivity_floor );
+  // eta = sqrt( SQ(resistivity_eta / eta_norm) + resistivity_floor*resistivity_floor ); // This mitigates discontinuities.
```
We can delete commented out code:
```suggestion
```

ZedThree reviewed (changes requested):
Some little bits to tidy up, but otherwise looks good. I'll leave the physics review up to @mikekryjak / @bendudson !

malamast committed:
parallel_ohms_law: Tidied up the code by removing unnecessary and unused methods.

malamast committed:
Removed some old renamimng conflicts related to electron_viscosity.

malamast committed:
parallel_ohms_law: Restractured the component so that it can be called before the sheath_boundary component.

using IS_SET_NOBOUNDARY and GET_NOBOUNDARY macros so that it can set Ve and NVe in the electron state without forcing it.
